expr/interval_compound_literal.go

```diff
@@ -0,0 +1,154 @@
+package expr
+
+import (
+	"errors"
+
+	"github.com/substrait-io/substrait-go/proto"
+	"github.com/substrait-io/substrait-go/types"
+)
+
+type intervalDateParts struct {
+	years      int32
+	months     int32
+	days       int32
+	seconds    int32
+	subSeconds int64
+}
+
+type intervalDatePartsOptions func(parts *intervalDateParts)
+
+func WithIntervalCompoundYears(years int32) func(*intervalDateParts) {
+	return func(idp *intervalDateParts) {
+		idp.years = years
+	}
+}
+
+func WithIntervalCompoundMonths(months int32) func(*intervalDateParts) {
+	return func(idp *intervalDateParts) {
+		idp.months = months
+	}
+}
+
+func WithIntervalCompoundDays(days int32) func(*intervalDateParts) {
+	return func(idp *intervalDateParts) {
+		idp.days = days
+	}
+}
+
+func WithIntervalCompoundSeconds(seconds int32) func(*intervalDateParts) {
+	return func(idp *intervalDateParts) {
+		idp.seconds = seconds
+	}
+}
+
+func WithIntervalCompoundSubSeconds(subSeconds int64) func(*intervalDateParts) {
+	return func(idp *intervalDateParts) {
+		idp.subSeconds = subSeconds
+	}
+}
```

Comment:
My thought was to keep it simple for user (set what is needed) and to not expose internal representation. Following reasons

1. I went to Option Pattern since there are more then two fields (more the fields then chances of mistakes are more) to set here and user will typically will set only some of them (depending on interval). If user needs to set Month they just need to give option "WithIntervalCompoundMonths" without worrying about other fields.

2. If we represent it as publicly exposed fields of "IntervalDatePartsConfig" and directly using it in our object. We would have to live with always taking this as input which would have made it hard to change internal representation in future. For internal representation to change, we would have to either copy these structure internally or made a breaking change.

Hence I felt option pattern is more maintainable and easier to use for user in this case

Reply:
Resolving this. Since it changed now to use builder pattern and public fields